Modules/_decimal/_decimal.c

```diff
@@ -2323,38 +2322,46 @@ static PyObject *
 dec_from_long(decimal_state *state, PyTypeObject *type, PyObject *v,
               const mpd_context_t *ctx, uint32_t *status)
 {
-    PyObject *dec;
-    PyLongObject *l = (PyLongObject *)v;
+    PyObject *dec = PyDecType_New(state, type);
 
-    dec = PyDecType_New(state, type);
     if (dec == NULL) {
         return NULL;
     }
 
-    if (_PyLong_IsZero(l)) {
-        _dec_settriple(dec, MPD_POS, 0, 0);
-        return dec;
-    }
-
-    uint8_t sign = _PyLong_IsNegative(l) ? MPD_NEG :  MPD_POS;
+    PyLongExport export_long;
 
-    if (_PyLong_IsCompact(l)) {
-        _dec_settriple(dec, sign, l->long_value.ob_digit[0], 0);
-        mpd_qfinalize(MPD(dec), ctx, status);
-        return dec;
+    if (PyLong_Export(v, &export_long) == -1) {
+        Py_DECREF(dec);
+        return NULL;
     }
-    size_t len = _PyLong_DigitCount(l);
+    if (export_long.digits) {
+        const PyLongLayout *layout = PyLong_GetNativeLayout();
+        const uint32_t base = (uint32_t)1 << layout->bits_per_digit;
+        const uint8_t sign = export_long.negative ? MPD_NEG : MPD_POS;
+        const Py_ssize_t len = export_long.ndigits;
 
-#if PYLONG_BITS_IN_DIGIT == 30
-    mpd_qimport_u32(MPD(dec), l->long_value.ob_digit, len, sign, PyLong_BASE,
-                    ctx, status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    mpd_qimport_u16(MPD(dec), l->long_value.ob_digit, len, sign, PyLong_BASE,
-                    ctx, status);
-#else
-  #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
+        if (base > UINT16_MAX) {
```

Comment:
We should use `digit_size` to chose between `mpd_qimport_u32` and `mpd_qimport_u16`. Values other than 2 or 4 are errors.

Reply:
You are assuming that `CHAR_BIT == 8`.  I'm not sure we should do this.

With added `assert(layout->bits_per_digit <= 32)` above (or RuntimeError) - I think nothing should be changed here.

Reply:
We are assuming that `CHAR_BIT == 8` everywhere in the code.

The new C API is used to untie the `decimal` code from the CPython implementation details. Other implementations can have `bits_per_digit == 8` and `digit_size == 1` or `bits_per_digit == 16` and `digit_size == 2`. While it is not particularly reasonable, but it is even possible to have `bits_per_digit < 16` and `digit_size > 2`. It is better to fail with a relevant error message than produce incorrect result or crash.

Doing the runtime check for every conversion can be too costly, so we can do this only at the module import (and cache the base value).

We should also check `digits_order` and `digit_endianness`. `mpd_qimport_u16` and `mpd_qimport_u32` only work if `digits_order` is -1 and `digit_endianness` is native. In other cases we should create a temporary digit array with expected order and endianness. Or at least raise a runtime error.